ql/src/java/org/apache/hadoop/hive/ql/exec/OperatorUtils.java

```diff
@@ -654,11 +655,18 @@ public static String findTableColNameOf(Operator<?> start, String internalColNam
         continue;
       }
       // If columnName is the output of a ColumnExpr get the original columnName from the Expr Map
-      if (currentOp.getColumnExprMap() != null && currentOp.getColumnExprMap().containsKey(internalColName)
-              && currentOp.getColumnExprMap().get(internalColName) instanceof ExprNodeColumnDesc) {
-        internalColName = ((ExprNodeColumnDesc) currentOp.getColumnExprMap().get(internalColName)).getColumn();
+      if (currentOp.getColumnExprMap() != null && currentOp.getColumnExprMap().containsKey(internalColName)) {
+        // Only use colInfo that is ExprNodeColumnDesc (could even be a UDF on the key at this point)
+        if (currentOp.getColumnExprMap().get(internalColName) instanceof ExprNodeColumnDesc) {
+          internalColName = ((ExprNodeColumnDesc) currentOp.getColumnExprMap().get(internalColName)).getColumn();
+          keyColInfo = currentOp.getSchema().getColumnInfo(internalColName);
```

Comment:
this doesn't look right to me; these 2 lines are mapping back to the previous ops colname and look it up on the current operator - if something like this happens - then the original bug should be fixed.

I think this fix may cause troubles later...

Reply:
this method somewhat reminds me to the `backtrack` methods we have in the `ExprNodeDescUtils` class - I don't know if you will find a perfect match ; but might worth to take a look at them

Reply:
Thanks for the comments @kgyrtkirk ! Updated the PR to make use of backtracking logic -- had to update it a bit to branch-out in a multi-parent scenario. Let me know what you think